pkg/flow/internal/controller/loader.go

```diff
@@ -402,7 +406,7 @@ func (l *Loader) populateServiceNodes(g *dag.Graph, serviceBlocks []*ast.BlockSt
 	for _, block := range serviceBlocks {
 		blockID := BlockComponentID(block).String()
 
-		if l.isModule() {
+		if !l.isRootController() {
```

Comment:
This loop is likely to panic now; line 419 tries to get a node in the graph that will never exist in non-root controllers, and then immediately calls `Block` on that nil node.

Reply:
there's a continue inside this block so I think we are ok. we did bump into this exact panic before getting the code right so I know what you mean.